add multiple appId,pull other appId config into config.

TerryLam2010 opened the pull request:
## What's the purpose of this PR
add multiple appId pull into config
see https://github.com/apolloconfig/apollo-java/issues/66

## Which issue(s) this PR fixes:
Fixes #


Follow this checklist to help us incorporate your contribution quickly and easily:

- [ ] add multiple appId pull into config.see https://github.com/apolloconfig/apollo-java/issues/66


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced `appId` parameter across various configuration classes, allowing configuration queries and changes to be associated with a specific application ID.
	- Added new method `getAccessKeySecret(String appId)` to manage access keys based on application IDs.
	- Implemented new methods for retrieving the default local cache directory using `appId`.
	- Added `MultipleConfig` annotation to support multiple application IDs.
	- Feature support for pulling configuration information from multiple AppIds.

- **Changes in APIs**
	- Updated method signatures in multiple configuration classes and services to include `appId`.
	- Enhanced configuration management methods to support application-specific identifiers.
	- Updated `ApolloConfigChangeListener` annotation to include `appId` attribute.
	- Modified `EnableApolloConfig` annotation to include `multipleConfigs` attribute for handling multiple configurations.

- **Refactor**
	- Refactored internal data structures and methods to support the inclusion of `appId`.

These changes enhance the ability to manage and query configurations on a per-application basis, improving flexibility and granularity in configuration management.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nobodyiam commented:
Thanks, it may take some time to reivew. In the meantime, could you please sign the CLA following the instructions provided by the CLA Assistant Lite bot?

TerryLam2010 commented:
I have read the CLA Document and I hereby sign the CLA

TerryLam2010 commented:
**_I have read the CLA Document and I hereby sign the CLA_**

TerryLam2010 commented:
**_I have read the CLA Document and I hereby sign the CLA_**

nobodyiam commented:
@TerryLam2010 It seems you committed the code with `zhantian.lin@watsons.com.cn`, which isn't associated with your GitHub account.

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/config/PropertySourcesProcessor.java:
```diff
@@ -55,22 +56,35 @@
+
   public static boolean addNamespaces(Collection<String> namespaces, int order) {
-    return NAMESPACE_NAMES.putAll(order, namespaces);
+    ConfigUtil configUtil = ApolloInjector.getInstance(ConfigUtil.class);
```
The file contains numerous instances of `configUtil` initialization, which isn't ideal.

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/property/AutoUpdateConfigChangeListener.java:
```diff
@@ -89,7 +89,7 @@ public void onChange(ConfigChangeEvent changeEvent) {
   }
 
-  private void updateSpringValue(SpringValue springValue) {
+  protected void updateSpringValue(SpringValue springValue) {
```
Why change the method modifiers?

TerryLam2010 replied:
AutoUpdateConfigChangeListener can be inherited by other classes, and some methods can be modified to protected, which can be used to implement your own logic. Some people can implement custom annotations and custom parsing, but event listening is required.

nobodyiam replied:
Since it’s not utilized anywhere in the project, I suggest we consider keeping it private.

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/property/SpringValue.java:
```diff
@@ -128,7 +128,7 @@ public boolean isJson() {
   }
 
-  boolean isTargetBeanValid() {
+  public boolean isTargetBeanValid() {
```
Why change the method modifiers?

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:
```diff
@@ -509,4 +509,11 @@ private boolean getPropertyBoolean(String propertyName, String envName, boolean
     return defaultVal;
   }
+
+  public String getAppSecret(String appId){
```
```suggestion
  public String getAccessKeySecret(String appId){
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:
```diff
@@ -509,4 +509,11 @@ private boolean getPropertyBoolean(String propertyName, String envName, boolean
+    if(appId.equals(getAppId())){
+      return getAccessKeySecret();
+    }
+    return System.getProperty("apollo.accesskey." + appId + ".secret", "");
```
It's better to add the logic to `Foundation.app().getAccessKeySecret`, which has a comprehensive list of places to check.

TerryLam2010 commented on apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfigManager.java:
```diff
@@ -24,47 +24,67 @@
 public class DefaultConfigManager implements ConfigManager {
   private ConfigFactoryManager m_factoryManager;
 
-  private Map<String, Config> m_configs = Maps.newConcurrentMap();
```
I don't think using string concatenation is a good approach for future expansion. Using Table<Row,Column,Value>, when you need to expand, you can retrieve all the configs under the appId.

TerryLam2010 commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/EnableApolloConfig.java:
```diff
@@ -69,4 +69,9 @@
+  /**
+   * Additional appId and namespace configurations. Will not participate in bootstrap
+   */
+  MultipleConfig[] multipleConfig() default {};
```
see com.ctrip.framework.apollo.spring.spi.DefaultApolloConfigRegistrarHelper#registerBeanDefinitions.The secret of @MultipleConfig is the key needed to load multiple appId configs.

nobodyiam replied:
```suggestion
  MultipleConfig[] multipleConfigs() default {};
```

Anilople commented:
The monitor metric #74  maybe changed if support multiple appId @Rawven

Rawven commented:
I will adapt my code after it is merged> The monitor metric #74  maybe changed if support multiple appId @Rawven

Rawven commented:
> 如果支持多个 appId，监控指标[#74可能会改变](https://github.com/apolloconfig/apollo-java/pull/74)@Rawven

I'll adapt it after it's merged

TerryLam2010 commented on apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java:
```diff
@@ -298,8 +306,8 @@ private Map<String, ConfigChange> updateAndCalcConfigChanges(Properties newConfi
-  private Properties loadFromResource(String namespace) {
-    String name = String.format("META-INF/config/%s.properties", namespace);
+  private Properties loadFromResource(String appId, String namespace) {
+    String name = String.format("META-INF/config/%s+%s.properties", appId, namespace);
```
All files have been changed to the app+namespace.properties format and there is no need to check the default appId.

nobodyiam replied:
By "breaking change," I mean that placing `somenamespace.properties` in the `META-INF/config` folder will no longer function after upgrading the apollo client.

TerryLam2010 commented:
I have changed the Email for submitting code to 36256759+TerryLam2010@users.noreply.github.com, and changed the name to TerryLam2010. Have not submitted open source request merge before,help me please....

TerryLam2010 commented:
**_I have read the CLA Document and I hereby sign the CLA_**

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/ConfigFile.java:
```diff
@@ -35,6 +35,14 @@ public interface ConfigFile {
    */
   boolean hasContent();
 
+  /**
+   * Get the appId of this config file instance
+   * @return the appId
+   */
+  default String getAppId(){
+    return null;
+  }
+
```
```suggestion
  /**
   * Get the appId of this config file instance
   * @return the appId
   *
   * @since 2.4.0
   */
  default String getAppId(){
    return null;
  }

```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/ApolloConfig.java:
```diff
@@ -49,6 +49,10 @@
 @Target(ElementType.FIELD)
 @Documented
 public @interface ApolloConfig {
+  /**
+   * Apollo appId for the config, if not specified then default to application
+   */
+  String appId() default "";
```
```suggestion
  /**
   * Apollo appId for the config, if not specified then default to the global appid
   *
   * @since 2.4.0
   */
  String appId() default "";
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/EnableApolloConfig.java:
```diff
@@ -69,4 +69,9 @@
    * @return
    */
   int order() default Ordered.LOWEST_PRECEDENCE;
+
+  /**
+   * Additional appId and namespace configurations. Will not participate in bootstrap
+   */
+  MultipleConfig[] multipleConfigs() default {};
```
```suggestion

  /**
   * Additional appId and namespace configurations. Will not participate in bootstrap
   *
   * @since 2.4.0
   */
  MultipleConfig[] multipleConfigs() default {};
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/MultipleConfig.java:
```diff
@@ -0,0 +1,45 @@
+
+import java.lang.annotation.*;
+
+/**
+ * @author Terry.Lam This annotation is used as a supplement to @EnableApolloConfig to fill in
+ * multiple appid
+ */
+
```
```suggestion
/**
 * @author Terry.Lam This annotation is used as a supplement to @EnableApolloConfig to fill in
 * multiple appid
 *
 * @since 2.4.0
 */
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/MultipleConfig.java:
```diff
@@ -0,0 +1,45 @@
+   * @return
+   */
+  String appId();
+
```
```suggestion
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/MultipleConfig.java:
```diff
@@ -0,0 +1,45 @@
+  String[] namespaces();
+
+  /**
+   * eg. ${apollo.multiple.shop.secret}
```
Please add more description to this field

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/config/PropertySourcesProcessor.java:
```diff
@@ -55,22 +56,32 @@
   private final ConfigPropertySourceFactory configPropertySourceFactory = SpringInjector
       .getInstance(ConfigPropertySourceFactory.class);
-  private ConfigUtil configUtil;
+  private static ConfigUtil configUtil = ApolloInjector.getInstance(ConfigUtil.class);
```
The initial logic is to set up the config util within the `postProcessBeanFactory` method. It's uncertain whether initializing it during the class initialization phase would be effective. For instance, the `apollo.cluster` could be defined in `application.properties` and subsequently assigned to system properties via `ApolloApplicationContextInitializer`.

TerryLam2010 replied:
This is a helpless operation. I have retained the default appId method of addNamespaces. addNamespaces is a static method. I need to declare it like this to get the value. Or does addNamespaces get an instance from ApolloInjector every time it is called?

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/internals/SimpleConfig.java:
```diff
@@ -37,18 +39,23 @@
   private volatile Properties m_configProperties;
   private volatile ConfigSourceType m_sourceType = ConfigSourceType.NONE;
 
+  private ConfigUtil m_configUtil = ApolloInjector.getInstance(ConfigUtil.class);
```
As there is a new `m_appId` field, there is no need to declare the `m_configUtil` as a global variable.

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/config/PropertySourcesProcessor.java:
```diff
@@ -55,22 +56,32 @@
   private ConfigurableEnvironment environment;
   private ApplicationEventPublisher applicationEventPublisher;
 
+
+  public PropertySourcesProcessor() {
+  }
+
```
Is this constructor necessary?

nobodyiam commented on apollo-core/src/main/java/com/ctrip/framework/foundation/spi/provider/ApplicationProvider.java:
```diff
@@ -46,4 +46,9 @@ public interface ApplicationProvider extends Provider {
+  /**
+   * @return the application's access key secret by appId
+   */
+  String getAccessKeySecret(String appId);
```
we could provide a default impl here

nobodyiam reviewed (commented):
Thank you for your hard work on this significant change. Here are a few comments for your review.

TerryLam2010 commented on apollo-client/src/main/java/com/ctrip/framework/apollo/internals/SimpleConfig.java:
```diff
@@ -111,7 +123,7 @@ public String apply(ConfigChange input) {
     clearConfigCache();
 
-    this.fireConfigChange(m_namespace, changeMap);
+    this.fireConfigChange(m_appId, m_namespace, changeMap);
```
Here use the m_appId of current object, rather than the method passed appId, is I imitate your m_namespace writing, determine to change to the passed appId?

nobodyiam replied:
Although the `appId` should be the same as the `m_appId`, I think it's better to use `appId` as it is passed in as a parameter.

TerryLam2010 commented on apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java:
```diff
@@ -61,8 +64,27 @@ public class DefaultConfig extends AbstractConfig implements RepositoryChangeLis
    * @param configRepository the config repository for this config instance
    */
   public DefaultConfig(String namespace, ConfigRepository configRepository) {
+    ConfigUtil configUtil = ApolloInjector.getInstance(ConfigUtil.class);
```
This is because the constructor this() must be the first line. I don't want to declare ConfigUtil as static, because non-static will show `Cannot reference 'DefaultConfig.configUtil' before supertype constructor has been called`. Do you have any suggestions?

nobodyiam replied:
We may pass appid as null and then assign it in the constructor:

```java
  public DefaultConfig(String appId, String namespace, ConfigRepository configRepository) {
    if (appId == null) {
        appId = ApolloInjector.getInstance(ConfigUtil.class).getAppId();
    }

```

nobodyiam commented:
> I have changed the Email for submitting code to [36256759+TerryLam2010@users.noreply.github.com](mailto:36256759+TerryLam2010@users.noreply.github.com), and changed the name to TerryLam2010. Have not submitted open source request merge before,help me please....

emm...You may first backup the local repo and try the following:
```bash
1. git checkout main
2. git checkout -b 20240703-apollo-multiple-appid-new
3. git merge --squash 20240703-apollo-multiple-appid
4. git commit -m "some commit message"
5. git push origin 20240703-apollo-multiple-appid-new:20240703-apollo-multiple-appid
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spi/DefaultConfigRegistry.java:
```diff
@@ -28,21 +33,37 @@
+
+  private ConfigUtil m_configUtil;
+
+  private Table<String,String,ConfigFactory> m_instances = Tables.synchronizedTable(HashBasedTable.create());
```
```suggestion
  private Table<String, String, ConfigFactory> m_instances = Tables.synchronizedTable(HashBasedTable.create());
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/ApolloConfig.java:
```diff
@@ -49,6 +49,12 @@
+   *
+   * @since 2.4.0
+   */
+  String appId() default "";
```
```suggestion
  String appId() default "";

```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/MultipleConfig.java:
```diff
@@ -0,0 +1,52 @@
+ *
+ * @since 2.4.0
+ */
+
```
```suggestion
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/config/PropertySourcesProcessor.java:
```diff
@@ -55,7 +56,7 @@
 public class PropertySourcesProcessor implements BeanFactoryPostProcessor, EnvironmentAware,
     ApplicationEventPublisherAware, PriorityOrdered {
-  private static final Multimap<Integer, String> NAMESPACE_NAMES = LinkedHashMultimap.create();
+  private static final Multimap<Integer, Multimap<String,String>> APP_NAMESPACE_NAMES = LinkedHashMultimap.create();
```
```suggestion
  private static final Multimap<Integer, Multimap<String, String>> APP_NAMESPACE_NAMES = LinkedHashMultimap.create();
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/config/PropertySourcesProcessor.java:
```diff
@@ -65,7 +66,14 @@ public class PropertySourcesProcessor implements BeanFactoryPostProcessor, Envir
 
   public static boolean addNamespaces(Collection<String> namespaces, int order) {
-    return NAMESPACE_NAMES.putAll(order, namespaces);
+    ConfigUtil configUtil = ApolloInjector.getInstance(ConfigUtil.class);
+    return addNamespaces(configUtil.getAppId(), namespaces, order);
```
```suggestion
    return addNamespaces(ApolloInjector.getInstance(ConfigUtil.class).getAppId(), namespaces, order);
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/config/PropertySourcesProcessor.java:
```diff
@@ -65,7 +66,14 @@ public class PropertySourcesProcessor implements BeanFactoryPostProcessor, Envir
+
+  public static boolean addNamespaces(String appId, Collection<String> namespaces, int order) {
+    Multimap<String,String> appNamespaceNames = LinkedHashMultimap.create();
+    appNamespaceNames.putAll(appId,namespaces);
```
```suggestion
    appNamespaceNames.putAll(appId, namespaces);
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/spi/DefaultApolloConfigRegistrarHelper.java:
```diff
@@ -47,8 +52,29 @@ public void registerBeanDefinitions(AnnotationMetadata importingClassMetadata, B
-    final String[] resolvedNamespaces = this.resolveNamespaces(namespaces);
-    PropertySourcesProcessor.addNamespaces(Lists.newArrayList(resolvedNamespaces), order);
+
+    Map<String,String[]> configMap = new HashMap<>();
```
```suggestion
    Map<String, String[]> configMap = new HashMap<>();
```

nobodyiam reviewed (changes requested):
This looks significantly better now. Please see some comments below:

TerryLam2010 commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spi/DefaultConfigFactoryManager.java:
```diff
@@ -29,12 +30,20 @@ public class DefaultConfigFactoryManager implements ConfigFactoryManager {
+  @Override
+  public ConfigFactory getFactory(String appId, String namespace) {
     // step 1: check hacked factory
     ConfigFactory factory = m_registry.getFactory(namespace);
```
Thank you, this is a major omissions!

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/config/PropertySourcesProcessor.java:
```diff
@@ -65,7 +66,13 @@
 
   public static boolean addNamespaces(Collection<String> namespaces, int order) {
-    return NAMESPACE_NAMES.putAll(order, namespaces);
+    return addNamespaces(ApolloInjector.getInstance(ConfigUtil.class).getAppId(), namespaces, order);
+  }
+
+  public static boolean addNamespaces(String appId, Collection<String> namespaces, int order) {
+    Multimap<String, String> appNamespaceNames = LinkedHashMultimap.create();
+    appNamespaceNames.putAll(appId, namespaces);
+    return APP_NAMESPACE_NAMES.put(order, appNamespaceNames);
```
This suggestion is valid; otherwise, we would override the app's namespaces.

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/internals/RemoteConfigLongPollService.java:
```diff
@@ -109,25 +107,34 @@ public RemoteConfigLongPollService() {
-    if (!m_longPollStarted.get()) {
-      startLongPolling();
+  public boolean submit(String appId, String namespace, RemoteConfigRepository remoteConfigRepository) {
+    Multimap<String, RemoteConfigRepository> repositoryMultimap = m_longPollNamespaces.get(
+        appId);
+    boolean result = false;
+    if(repositoryMultimap == null){
+      repositoryMultimap = Multimaps.synchronizedSetMultimap(HashMultimap.<String, RemoteConfigRepository>create());
+      result = repositoryMultimap.put(namespace, remoteConfigRepository);
+      m_longPollNamespaces.put(appId, repositoryMultimap);
+    }else{
+      result = repositoryMultimap.put(namespace, remoteConfigRepository);
     }
-    return added;
+    m_notifications.put(appId, namespace, INIT_NOTIFICATION_ID);
+    if (m_longPollStarted.get(appId) == null) {
+      startLongPolling(appId);
+    }
+    return result;
```
This is a valid suggestion.

nobodyiam commented on apollo-client/src/test/java/com/ctrip/framework/apollo/internals/RemoteConfigLongPollServiceTest.java:
```diff
@@ -90,13 +100,16 @@ public void setUp() throws Exception {
 
     MockInjector.setInstance(ConfigUtil.class, new MockConfigUtil());
 
+/*    Field field = PropertySourcesProcessor.class.getDeclaredField("configUtil");
+    field.setAccessible(true);
+    field.set(null, ApolloInjector.getInstance(ConfigUtil.class));*/
+
```
Why are these lines commented out?

TerryLam2010 replied:
This deprecated code has been removed.

nobodyiam commented on apollo-client/src/test/java/com/ctrip/framework/apollo/internals/RemoteConfigRepositoryTest.java:
```diff
@@ -101,6 +103,10 @@ public void setUp() throws Exception {
     configUtil = new MockConfigUtil();
     MockInjector.setInstance(ConfigUtil.class, configUtil);
 
+/*    Field field = PropertySourcesProcessor.class.getDeclaredField("configUtil");
```
The same as above.

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/internals/RemoteConfigLongPollService.java:
```diff
@@ -289,8 +298,8 @@ private void updateRemoteNotifications(List<ApolloConfigNotification> deltaNotif
 
-  private String assembleNamespaces() {
-    return STRING_JOINER.join(m_longPollNamespaces.keySet());
+  private String assembleNamespaces(String appId) {
+    return STRING_JOINER.join(m_longPollNamespaces.get(appId).keySet());
```
It's better to add a check.

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/internals/RemoteConfigLongPollService.java:
```diff
@@ -226,20 +235,20 @@
-  private void notify(ServiceDTO lastServiceDto, List<ApolloConfigNotification> notifications) {
+  private void notify(String appId, ServiceDTO lastServiceDto, List<ApolloConfigNotification> notifications) {
     if (notifications == null || notifications.isEmpty()) {
       return;
     }
     for (ApolloConfigNotification notification : notifications) {
       String namespaceName = notification.getNamespaceName();
       //create a new list to avoid ConcurrentModificationException
       List<RemoteConfigRepository> toBeNotified =
-          Lists.newArrayList(m_longPollNamespaces.get(namespaceName));
+          Lists.newArrayList(m_longPollNamespaces.get(appId).get(namespaceName));
       ApolloNotificationMessages originalMessages = m_remoteNotificationMessages.get(namespaceName);
       ApolloNotificationMessages remoteMessages = originalMessages == null ? null : originalMessages.clone();
       //since .properties are filtered out by default, so we need to check if there is any listener for it
       toBeNotified.addAll(m_longPollNamespaces
-          .get(String.format("%s.%s", namespaceName, ConfigFileFormat.Properties.getValue())));
+          .get(appId).get(String.format("%s.%s", namespaceName, ConfigFileFormat.Properties.getValue())));
```
This suggestion is valid.

nobodyiam reviewed (changes requested):
Thank you for your hard work on the pull request; I believe it's nearing approval. Please see my comments below.

nobodyiam commented:
> > I have changed the Email for submitting code to [36256759+TerryLam2010@users.noreply.github.com](mailto:36256759+TerryLam2010@users.noreply.github.com), and changed the name to TerryLam2010. Have not submitted open source request merge before,help me please....
> 
> emm...You may first backup the local repo and try the following:
> 
> ```shell
> 1. git checkout main
> 2. git checkout -b 20240703-apollo-multiple-appid-new
> 3. git merge --squash 20240703-apollo-multiple-appid
> 4. git commit -m "some commit message"
> 5. git push origin 20240703-apollo-multiple-appid-new:20240703-apollo-multiple-appid
> ```

I noticed the CLA check failed again, and it seems the original commits are back. Is there some mistaken operation on the branch?

![image](https://github.com/user-attachments/assets/c92f57ee-8771-43ed-ad11-61dc0f864e46)

TerryLam2010 commented on apollo-client/src/test/java/com/ctrip/framework/apollo/ConfigServiceTest.java:
```diff
@@ -43,8 +45,12 @@ public class ConfigServiceTest {
     someAppId = "someAppId";
-
     MockInjector.setInstance(ConfigUtil.class, new MockConfigUtil());
+/*    Field field = ConfigService.class.getDeclaredField("m_configUtil");
+    field.setAccessible(true);
+    MockInjector.setInstance(ConfigUtil.class, new MockConfigUtil());
+    field.set(null, ApolloInjector.getInstance(ConfigUtil.class));*/
+
```
Restored to your version

TerryLam2010 commented on apollo-client/src/test/java/com/ctrip/framework/apollo/build/MockInjector.java:
```diff
@@ -73,7 +74,8 @@ public static class InjectCustomizer implements ApolloInjectorCustomizer {
     @Override
     public <T> T getInstance(Class<T> clazz) {
-      return (T) classMap.get(clazz);
+      T instance = (T) classMap.get(clazz);
+      return instance;
```
Restored to your version

TerryLam2010 commented on apollo-client/src/test/java/com/ctrip/framework/apollo/integration/ConfigIntegrationTest.java:
```diff
@@ -371,7 +371,7 @@ public void onChange(ConfigChangeEvent changeEvent) {
     );
 
-    longPollFinished.get(5000, TimeUnit.MILLISECONDS);
+    longPollFinished.get(10000, TimeUnit.MILLISECONDS);
```
Sometimes my unit test cannot pass when set to 5000, but the pass rate is high when set to 10000. I will change it back.

TerryLam2010 commented on apollo-client/src/test/java/com/ctrip/framework/apollo/internals/RemoteConfigLongPollServiceTest.java:
```diff
@@ -80,6 +84,12 @@ public class RemoteConfigLongPollServiceTest {
+    someAppId = "someAppId";
+    someCluster = "someCluster";
+
+    System.setProperty(ApolloClientSystemConsts.APP_ID, someAppId);
```
deleted System.setProperty(ApolloClientSystemConsts.APP_ID, someAppId);

nobodyiam commented on apollo-client/src/test/java/com/ctrip/framework/apollo/internals/RemoteConfigLongPollServiceTest.java:
```diff
@@ -28,11 +28,14 @@
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
+import com.ctrip.framework.apollo.build.ApolloInjector;
 import com.ctrip.framework.apollo.build.MockInjector;
+import com.ctrip.framework.apollo.core.ApolloClientSystemConsts;
```
Good point.

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/internals/RemoteConfigLongPollService.java:
```diff
@@ -109,25 +111,31 @@ public RemoteConfigLongPollService() {
-    if (!m_longPollStarted.get()) {
-      startLongPolling();
+  public boolean submit(String appId, String namespace, RemoteConfigRepository remoteConfigRepository) {
+    Multimap<String, RemoteConfigRepository> repositoryMultimap = m_longPollNamespaces.get(appId);
+    boolean result = false;
+    if(repositoryMultimap == null){
+      repositoryMultimap = Multimaps.synchronizedSetMultimap(HashMultimap.<String, RemoteConfigRepository>create());
+      m_longPollNamespaces.put(appId, repositoryMultimap);
```
Nice suggestion!

nobodyiam commented on apollo-client/src/test/java/com/ctrip/framework/apollo/spring/config/CachedCompositePropertySourceTest.java:
```diff
@@ -82,7 +82,7 @@ public void testGetPropertyNames() {
     assertSame(returnedPropertyNames, compositeSource.getPropertyNames());
 
-    listeners.get(0).onChange(new ConfigChangeEvent(null, null));
+    listeners.get(0).onChange(new ConfigChangeEvent(null,null, null));
```
The constructor of `ConfigChangeEvent` was changed to accept three parameters.

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spi/DefaultConfigFactoryManager.java:
```diff
@@ -27,23 +30,31 @@
   private ConfigRegistry m_registry;
 
-  private Map<String, ConfigFactory> m_factories = Maps.newConcurrentMap();
+  private Table<String, String, ConfigFactory> m_factories = HashBasedTable.create();
```
This is a valid point.

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfigManager.java:
```diff
@@ -24,59 +24,80 @@
-  private Map<String, Config> m_configs = Maps.newConcurrentMap();
+  private ConfigUtil m_configUtil;
+
+  private Table<String,String,Config> m_configs = Tables.synchronizedTable(HashBasedTable.<String,String,Config>create());
```
This is a valid point.

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/ConfigService.java:
```diff
@@ -93,16 +99,32 @@
 
+      @Override
+      public Config create(String appId, String namespace) {
+        if(!StringUtils.equals(appId, m_configUtil.getAppId())){
+          throw new IllegalArgumentException("appId not match");
+        }
```
This is a valid point.

nobodyiam commented on apollo-client/src/test/java/com/ctrip/framework/apollo/ConfigServiceTest.java:
```diff
@@ -151,6 +162,11 @@ public boolean hasContent() {
       return true;
     }
 
+    @Override
+    public String getAppId() {
+      return null;
+    }
```
This is a valid point.

nobodyiam commented on apollo-client/src/test/java/com/ctrip/framework/apollo/spring/AbstractSpringIntegrationTest.java:
```diff
@@ -161,15 +167,21 @@ protected Date assembleDate(int year, int month, int day, int hour, int minute,
+  }
+
+  protected static void doSetUp() throws Exception {
+    System.setProperty(ApolloClientSystemConsts.APP_ID, "someAppId");
```
It's better to use mock instead of setting the system property.

nobodyiam commented on apollo-client/src/test/java/com/ctrip/framework/apollo/spring/AbstractSpringIntegrationTest.java:
```diff
@@ -178,9 +190,13 @@ protected static void doSetUp() {
   }
 
-  protected static void doTearDown() {
+  protected static void doTearDown(){
```
```suggestion
  protected static void doTearDown() {
```

nobodyiam commented on apollo-client/src/test/java/com/ctrip/framework/apollo/spring/AbstractSpringIntegrationTest.java:
```diff
@@ -195,16 +211,26 @@ public MockConfigManager(ConfigManager delegate) {
     @Override
     public Config getConfig(String namespace) {
-      Config config = CONFIG_REGISTRY.get(namespace);
+      return getConfig("someAppId", namespace);
```
Do not hard code the appid or it's very hard to refactor in the future.

nobodyiam commented on apollo-client/src/test/java/com/ctrip/framework/apollo/spring/AbstractSpringIntegrationTest.java:
```diff
@@ -178,9 +190,13 @@
   }
 
-  protected static void doTearDown() {
+  protected static void doTearDown(){
     MockInjector.reset();
     CONFIG_REGISTRY.clear();
```
good point.

nobodyiam commented on apollo-client/src/test/java/com/ctrip/framework/apollo/spring/AbstractSpringIntegrationTest.java:
```diff
@@ -85,14 +91,14 @@
-  protected SimpleConfig prepareConfig(String namespaceName, Properties properties) {
+  protected SimpleConfig prepareConfig(String appId, String namespaceName, Properties properties) {
     ConfigRepository configRepository = mock(ConfigRepository.class);
 
     when(configRepository.getConfig()).thenReturn(properties);
 
-    SimpleConfig config = new SimpleConfig(ConfigConsts.NAMESPACE_APPLICATION, configRepository);
+    SimpleConfig config = new SimpleConfig(appId, ConfigConsts.NAMESPACE_APPLICATION, configRepository);
 
-    mockConfig(namespaceName, config);
+    mockConfig(appId, namespaceName, config);
```
This issue is not introduced in this pull request but it's valuable to fix it.

nobodyiam reviewed (commented):
Please find some more comments as below.

TerryLam2010 commented on apollo-client/src/main/java/com/ctrip/framework/apollo/internals/SimpleConfig.java:
```diff
@@ -50,6 +53,21 @@ public class SimpleConfig extends AbstractConfig implements RepositoryChangeList
+   */
+  public SimpleConfig(String appId, String namespace, ConfigRepository configRepository) {
+    if (appId == null) {
+      appId = ApolloInjector.getInstance(ConfigUtil.class).getAppId();
+    }
```
It is necessary to return a null, and the null pointer judgment user will judge

TerryLam2010 commented on apollo-client/src/main/java/com/ctrip/framework/apollo/internals/LocalFileConfigRepository.java:
```diff
@@ -63,11 +64,12 @@
    */
-  public LocalFileConfigRepository(String namespace) {
-    this(namespace, null);
+  public LocalFileConfigRepository(String appId, String namespace) {
+    this(appId, namespace, null);
```
The appId must be passed in from the upper method, so there is no problem here

TerryLam2010 commented:
Jason, my unit tests pass perfectly when I run them locally, but fail when I run them on GitHub.

nobodyiam commented on apollo-client/src/test/java/com/ctrip/framework/apollo/internals/RemoteConfigLongPollServiceTest.java:
```diff
@@ -28,11 +28,14 @@
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
+import com.ctrip.framework.apollo.build.ApolloInjector;
```
I believe these imports are not used.

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spi/DefaultConfigFactoryManager.java:
```diff
@@ -16,34 +16,43 @@
+    return getFactory(m_configUtil.getAppId(), namespace);
+  }
+
+  @Override
+  public ConfigFactory getFactory(String appId, String namespace) {
```
There is a getFactory(String appId, String namespace) method in the ConfigFactoryManager interface.

nobodyiam reviewed (commented):
I ran the unit test in my local environment and get the same errors.

```
[ERROR] Failures: 
[ERROR]   ConfigIntegrationTest.testGetConfigWithLocalFileAndRemoteConfigError:179 expected: <someValue> but was: <anotherValue>
[ERROR]   ConfigIntegrationTest.testGetConfigWithNoLocalFileAndRemoteConfigError:164 expected: <defaultValue0.4766152787709471> but was: <someValue>
[ERROR]   ConfigIntegrationTest.testOrderGetConfigWithLocalFileAndRemoteConfigError:201 expected: <someValue1> but was: <null>
[ERROR]   ConfigServiceTimeoutIntegrationTest.give9NamespaceConfigBackupWhenMeetTimeout:156 expected: <backupNs0.v1> but was: <null>
[ERROR]   ConfigServiceTimeoutIntegrationTest.giveDefaultNamespaceConfigBackupWhenMeetTimeout:96 expected: <v1> but was: <null>
[ERROR]   ConfigServiceTimeoutIntegrationTest.giveTimeoutThenRecover:198 expected: <v1> but was: <null>
[ERROR]   ConfigServiceLocatorTest.giveNoUsableConfigServiceThenThrowExceptionQuickly:84 Expected com.ctrip.framework.apollo.exceptions.ApolloConfigException to be thrown, but nothing was thrown.
[INFO] 
[ERROR] Tests run: 428, Failures: 7, Errors: 0, Skipped: 1
```

nobodyiam commented:
@TerryLam2010 As this pull request is quite large and already takes a lot of time in reviewing. I suggest scheduling an online meeting for a thorough, quick review. If that works for you, please email me at [nobodyiam@gmail.com](mailto:nobodyiam@gmail.com) to arrange a suitable time.

TerryLam2010 commented on apollo-client/src/test/java/com/ctrip/framework/apollo/spring/config/PropertySourcesProcessorTest.java:
```diff
@@ -51,6 +58,7 @@
   @Override
   @Before
   public void setUp() throws Exception {
+    System.setProperty(ApolloClientSystemConsts.APP_ID, someAppId);
```
remove it.

TerryLam2010 replied:
I will remove this code.

TerryLam2010 commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/spi/DefaultApolloConfigRegistrarHelper.java:
```diff
@@ -47,8 +52,32 @@
+        String appId = multipleConfig.getString("appId");
+        String[] multipleNamespaces = this.resolveNamespaces(multipleConfig.getStringArray("namespaces"));
+        configMap.put(appId, multipleNamespaces);
+        String secret = resolveSecret(multipleConfig.getString("secret"));
+        int multipleOrder = multipleConfig.getNumber("order");
+
+        // put multiple secret into system property
+        System.setProperty("apollo.accesskey." + appId + ".secret", secret);
```
Allow null.Judging the empty place where it is used

TerryLam2010 commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/spi/DefaultApolloConfigRegistrarHelper.java:
```diff
@@ -77,6 +106,16 @@
     return resolvedNamespaces;
   }
 
+  private String resolveSecret(String secret){
+    if (this.environment == null) {
+      if (secret.contains("${")) {
+        logger.warn("secret placeholder {} is not supported for Spring version prior to 3.2.x", secret);
+      }
+      return secret;
+    }
+    return this.environment.resolveRequiredPlaceholders(secret);
+  }
```
This is good suggestion

nobodyiam commented on apollo-client/src/test/java/com/ctrip/framework/apollo/spring/XmlConfigPlaceholderTest.java:
```diff
@@ -62,15 +62,15 @@ public void testPropertySourceWithNoNamespace() throws Exception {
     when(config.getProperty(eq(BATCH_PROPERTY), Mockito.nullable(String.class))).thenReturn(String.valueOf(someBatch));
 
-    mockConfig(ConfigConsts.NAMESPACE_APPLICATION, config);
+    mockConfig(someAppId, ConfigConsts.NAMESPACE_APPLICATION, config);
```
It is defined in the parent class AbstractSpringIntegrationTest.

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/internals/RemoteConfigLongPollService.java:
```diff
@@ -255,20 +263,20 @@
     for (ApolloConfigNotification notification : deltaNotifications) {
       if (Strings.isNullOrEmpty(notification.getNamespaceName())) {
         continue;
       }
       String namespaceName = notification.getNamespaceName();
-      if (m_notifications.containsKey(namespaceName)) {
-        m_notifications.put(namespaceName, notification.getNotificationId());
+      if (m_notifications.contains(appId, namespaceName)) {
+        m_notifications.put(appId, namespaceName, notification.getNotificationId());
       }
       //since .properties are filtered out by default, so we need to check if there is notification with .properties suffix
       String namespaceNameWithPropertiesSuffix =
           String.format("%s.%s", namespaceName, ConfigFileFormat.Properties.getValue());
-      if (m_notifications.containsKey(namespaceNameWithPropertiesSuffix)) {
-        m_notifications.put(namespaceNameWithPropertiesSuffix, notification.getNotificationId());
+      if (m_notifications.contains(appId, namespaceNameWithPropertiesSuffix)) {
+        m_notifications.put(appId, namespaceNameWithPropertiesSuffix, notification.getNotificationId());
```
The m_notifications is a synchronizedTable, so I think it won't have potential race condition issue?

nobodyiam reviewed (commented):
Great work! It's ready to merge. Please address the minor comments below and update CHANGES.md.

nobodyiam commented on CHANGES.md:
```diff
@@ -10,6 +10,9 @@ Apollo Java 2.4.0
 * [Add more observability in apollo config client](https://github.com/apolloconfig/apollo-java/pull/74)
 * [Feature Support Kubernetes ConfigMap cache for Apollo java, golang client](https://github.com/apolloconfig/apollo-java/pull/79)
+* [Feature support pulling configuration information from multiple AppIds](https://github.com/apolloconfig/apollo-java/pull/70)
+
+
```
```suggestion
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/EnableApolloConfig.java:
```diff
@@ -69,4 +69,13 @@
+  /**
+   * Additional appId and namespace configurations. Will not participate in bootstrap
+   * e.g.
+   *   @EnableApolloConfig(value = {"mainAppId", "APPLICATION_YAML_NAMESPACE"},
+   *       multipleConfigs = {@MultipleConfig(appId = "multipleAppId", namespaces = {"namespace2"}, secert="${xxxx.secret}")})
```
```suggestion
   *   @EnableApolloConfig(value = {"someNamespace", "anotherNamespace"},
   *       multipleConfigs = {@MultipleConfig(appId = "anotherAppId", namespaces = {"yetAnotherNamespace"}, secert="${xxxx.secret}")})
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/spi/DefaultApolloConfigRegistrarHelper.java:
```diff
@@ -47,8 +52,26 @@ public void registerBeanDefinitions(AnnotationMetadata importingClassMetadata, B
+        int multipleOrder = multipleConfig.getNumber("order");
+
+        // put multiple secret into system property
+        if(secret != null){
```
```suggestion
        if (!StringUtils.isBlank(secret)) {
```

nobodyiam commented on apollo-client/src/main/java/com/ctrip/framework/apollo/spring/spi/DefaultApolloConfigRegistrarHelper.java:
```diff
@@ -16,6 +16,7 @@
 package com.ctrip.framework.apollo.spring.spi;
 
+import com.ctrip.framework.apollo.build.ApolloInjector;
 import com.ctrip.framework.apollo.core.spi.Ordered;
```
```suggestion
import com.ctrip.framework.apollo.core.spi.Ordered;
import com.ctrip.framework.apollo.core.utils.StringUtils;
```

nobodyiam reviewed (approved):
LGTM
